kernel/wait_queue/src/lib.rs

```diff
@@ -1,239 +1,65 @@
 #![allow(clippy::new_without_default)]
 #![no_std]
 
-extern crate alloc;
-#[macro_use] extern crate log;
-extern crate irq_safety;
-extern crate task;
-extern crate scheduler;
-
-
-use alloc::collections::VecDeque;
-use irq_safety::MutexIrqSafe;
-use task::{TaskRef, RunState};
-
-
-/// An object that holds a blocked `Task` 
-/// that will be automatically unblocked upon drop.  
-pub struct WaitGuard {
-    task: TaskRef,
-}
-impl WaitGuard {
-    /// Blocks the given `Task` and returns a new `WaitGuard` object
-    /// that will automatically unblock that Task when it is dropped. 
-    ///
-    /// Returns an error if the task cannot be blocked;
-    /// see [`task::Task::block()`] for more details.
-    pub fn new(task: TaskRef) -> Result<WaitGuard, RunState> {
-        task.block()?;
-        Ok(WaitGuard { task })
-    }
-
-    /// Blocks the task guarded by this waitguard,
-    /// which is useful to re-block a task after it spuriously woke up. 
-    ///
-    /// Returns an error if the task cannot be blocked;
-    /// see [`task::Task::block()`] for more details.
-    pub fn block_again(&self) -> Result<RunState, RunState> {
-        self.task.block()
-    }
-
-    /// Returns a reference to the `Task` being blocked in this `WaitGuard`.
-    pub fn task(&self) -> &TaskRef {
-        &self.task
-    }
-}
-impl Drop for WaitGuard {
-    fn drop(&mut self) {
-        self.task.unblock().unwrap();
-    }
-}
-
-
-/// Errors that may occur while waiting on a waitqueue/condition/event.
-#[derive(Debug, PartialEq)]
-pub enum WaitError {
-    NoCurrentTask,
-    Interrupted,
-    Timeout,
-    SpuriousWakeup,
-    CantBlockCurrentTask,
+use mpmc_queue::Queue;
+use sync::DeadlockPrevention;
+use sync_spin::Spin;
+use task::{get_my_current_task, TaskRef};
+
+pub struct WaitQueue<P = Spin>
```

Comment:
if a waitqueue only has the semantics of blocking a task, then why do we need to parameterize it on any form of `DeadlockPrevention`? It seems like this struct should be the one to make the choice as to what we're doing here, which AFAICT is just blocking a task.